data_lang/pretty.py

```diff
@@ -1,40 +1,546 @@
+#
+# ~ Constructors ~
+#
+# There are just a few constructors for `doc`, from which everything else is
+# built from.
+#
+# Text(string) prints like:
+# |string
+#
+# Break(string) prints like:
+# |string
+# or like a newline:
+# |
+# |
+# (It does the latter if printed in "flat" mode, and the former otherwise. See
+# Group for details.)
+#
+# Concat(a, b) prints like:
+# |AAAAA
+# |AAABBB
+# |BBBBB
+#
+# Indent(3, a) prints like:
+# |AAAAA
+# |   AAAAA
+# |   AAAAA
+# (notice that the first line isn't indented)
+#
+# Group(a) makes a decision. It either:
+# - Prints `a` "flat", meaning that (i) every Break inside of it is printed as a
+#   string instead of as a newline, and (ii) every Group nested inside of it is
+#   printed flat.
+# - Prints `a` normally, meaning that (i) the Breaks inside of it are printed as
+#   newlines, and (ii) the Groups inside of it make their own decision about
+#   whether to be flat.
+# It makes this decision greedily. If the current line would not overflow if the
+# group printed flat, then it will print flat. This takes into account not only
+# the group itself, but the content before and after it on the same line.
+#
+# ~ Measures ~
+#
+# The algorithm used here is close to the one originally described by Wadler,
+# but it precomputes a "measure" for each node in the `doc`. This "measure"
+# allows each Groups to decide whether to print flat or not without needing to
+# look ahead per the standard algorithm. A measure has two pieces of
+# information:
+#
+# - Measure.flat is the width of the doc if it's printed flat.
+# - Measure.nonflat is the width of the doc until the _earliest possible_
+#   newline, or -1 if it doesn't contain a Break.
+#
+# Measures are used in two steps. First, they're computed bottom-up on the
+# `doc`, measuring the size of each node. Later, _PrintDoc() stores a measure in
+# each DocFragment. These Measures measure something different: the width from
+# the doc _to the end of the entire doc tree_. This second set of Measures (the
+# ones in the DocFragments) are computed top-down, and they're used to decide
+# for each Group whether to use flat mode or not, without needing to scan ahead.
+
 from __future__ import print_function
 
-from _devbuild.gen.pretty_asdl import doc, doc_t
-from _devbuild.gen.value_asdl import value_e, value_t
+from _devbuild.gen.pretty_asdl import doc, doc_e, DocFragment, Measure, MeasuredDoc
+from _devbuild.gen.value_asdl import value, value_e, value_t, value_str
+
+from data_lang.j8 import ValueIdString, HeapValueId
+from typing import cast, List, Dict
+from core import ansi
+from libc import wcswidth
+from frontend import match
 
-from mycpp.mylib import log, tagswitch
+import fastfunc
+
+from mycpp import mops
+from mycpp.mylib import log, tagswitch, BufWriter, iteritems
 
 _ = log
 
+KEY_STYLE = ansi.GREEN
+NUMBER_STYLE = ansi.YELLOW
+NULL_STYLE = ansi.BOLD + ansi.RED
+BOOL_STYLE = ansi.BOLD + ansi.BLUE
+CYCLE_STYLE = ansi.BOLD + ansi.YELLOW
```

Comment:
These styles (and the choice of which things are styled) can of course be whatever you like. I picked these out of a hat. Not sure if there's anything more principled to do. Could copy nodejs styling, I suppose.